### PR TITLE
storage: fix disappearing RHS merge bug

### LINKED ISSUE #31719: ccl/importccl: TestImportCSVStmt failed under stress
SHA: https://github.com/cockroachdb/cockroach/commits/d3f39a5cb60eb9cea64635f78e25f67a435057a5

Parameters:
```
TAGS=
GOFLAGS=
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
make stressrace TESTS=TestImportCSVStmt PKG=github.com/cockroachdb/cockroach/pkg/ccl/importccl TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=980299&tab=buildLog

```

I181023 05:34:49.586318 33503 storage/store_snapshot.go:636  [n1,raftsnapshot,s1,r30/1:/Table/75/2/"{J"/4326-N"/1808}] sending Raft snapshot b2c8df3b at applied index 19
I181023 05:34:49.588054 33503 storage/store_snapshot.go:679  [n1,raftsnapshot,s1,r30/1:/Table/75/2/"{J"/4326-N"/1808}] streamed snapshot to (n3,s3):2: kv pairs: 684, log entries: 9, rate-limit: 8.0 MiB/sec, 2ms
I181023 05:34:49.594509 33506 storage/replica_raftstorage.go:807  [n3,s3,r30/2:{-}] applying Raft snapshot at index 19 (id=b2c8df3b, encoded size=31478, 1 rocksdb batches, 9 log entries)
I181023 05:34:49.595766 33506 storage/replica_raftstorage.go:813  [n3,s3,r30/2:/Table/75/2/"{J"/4326-N"/1808}] applied Raft snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
I181023 05:34:49.609580 9149 sql/event_log.go:126  [n1,client=127.0.0.1:60632,user=root] Event: "create_database", target: 81, info: {DatabaseName:csv19 Statement:CREATE DATABASE csv19 User:root}
I181023 05:34:49.616381 8021 storage/store.go:2580  [n1,s1,r29/1:/Table/75/2/"{G"/1931-J"/4326}] removing replica r30/1
I181023 05:34:49.617863 8518 storage/store.go:2580  [n2,s2,r29/3:/Table/75/2/"{G"/1931-J"/4326}] removing replica r30/3
I181023 05:34:49.619024 33294 storage/store_snapshot.go:636  [n1,raftsnapshot,s1,r29/1:/Table/75/2/"{G"/1931-N"/1808}] sending Raft snapshot ed24128a at applied index 19
I181023 05:34:49.620528 33294 storage/store_snapshot.go:679  [n1,raftsnapshot,s1,r29/1:/Table/75/2/"{G"/1931-N"/1808}] streamed snapshot to (n3,s3):2: kv pairs: 684, log entries: 9, rate-limit: 8.0 MiB/sec, 12ms
I181023 05:34:49.620976 33605 storage/replica_raftstorage.go:807  [n3,s3,r29/2:{-}] applying Raft snapshot at index 19 (id=ed24128a, encoded size=31954, 1 rocksdb batches, 9 log entries)
I181023 05:34:49.622244 33575 storage/store.go:2580  [n3,replicaGC,s3,r30/2:/Table/75/2/"{J"/4326-N"/1808}] removing replica r30/2
I181023 05:34:49.625133 33575 storage/replica.go:863  [n3,replicaGC,s3,r30/2:/Table/75/2/"{J"/4326-N"/1808}] removed 680 (671+9) keys in 0ms [clear=0ms commit=0ms]
I181023 05:34:49.622765 33605 storage/replica_raftstorage.go:813  [n3,s3,r29/2:/Table/75/2/"{G"/1931-J"/4326}] applied Raft snapshot in 1ms [clear=0ms batch=0ms entries=0ms commit=0ms]
F181023 05:34:49.657430 8751 storage/replica.go:5713  [n3,s3,r29/2:/Table/75/2/"{G"/1931-J"/4326}] RHS of merge r29:/Table/75/2/"{G"/1931-N"/1808} [(n1,s1):1, (n3,s3):2, (n2,s2):3, next=4, gen=2] <- r30:/Table/75/2/"{J"/4326-N"/1808} [(n1,s1):1, (n3,s3):2, (n2,s2):3, next=4, gen=1] not present on store; found r30:{-} [<no replicas>, next=0, gen=0] in place of the RHS
goroutine 8751 [running]:
github.com/cockroachdb/cockroach/pkg/util/log.getStacks(0xc4200c6e01, 0xc4200c6ea0, 0x3eb4500, 0x12)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:997 +0xcf
github.com/cockroachdb/cockroach/pkg/util/log.(*loggingT).outputLogEntry(0x44763a0, 0xc400000004, 0x3eb45d0, 0x12, 0x1651, 0xc4256d4c80, 0x13b)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/clog.go:866 +0x8e1
github.com/cockroachdb/cockroach/pkg/util/log.addStructured(0x303b040, 0xc425515aa0, 0x4, 0x2, 0x2b96eaf, 0x48, 0xc423d969f8, 0x3, 0x3)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/structured.go:85 +0x2e5
github.com/cockroachdb/cockroach/pkg/util/log.logDepth(0x303b040, 0xc425515aa0, 0x1, 0x4, 0x2b96eaf, 0x48, 0xc423d969f8, 0x3, 0x3)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:69 +0x8c
github.com/cockroachdb/cockroach/pkg/util/log.Fatalf(0x303b040, 0xc425515aa0, 0x2b96eaf, 0x48, 0xc423d969f8, 0x3, 0x3)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/log.go:172 +0x7e
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).acquireMergeLock(0xc422374f00, 0x303b040, 0xc425515aa0, 0xc424fe3e00, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5713 +0x1e5
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).maybeAcquireSplitMergeLock(0xc422374f00, 0x303b040, 0xc425515aa0, 0x100000001, 0x1, 0x5, 0x0, 0x7, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5648 +0x66
```

### PR DESCRIPTION
**Pull request by benesch**
The strategy used by the replica GC queue to determine whether a
subsumed range can be GC'd is flawed. If a replica of the LHS was
uninitialized at the time the merge commmitted, there is a small window
where the replica GC queue can think that it's safe to clean up an RHS
replica when in fact the uninitialized LHS replica could still
initialize and apply a merge trigger that required that RHS to be
present.

Make the replica GC queue's strategy valid by requiring that all
replicas of the LHS are initialized before beginning a merge
transaction. This closes the window during which a replica of the RHS
could be incorrectly GC'd with a patch that is small enough to be
backported to v2.1.1.

Fix #31719.

Release note: None

### CODE REVIEW

**Comment by benesch** [comment]
Ok, this is RFAL.

**Review by tbg — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31842)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/storage/replica_command.go, line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPeTjB0-EH3WhpkwgVF:-LPeTjB0-EH3WhpkwgVG:bhnssts) ([raw file](https://github.com/cockroachdb/cockroach/blob/7578c501755025d8fe4dff109e452871ca3e9432/pkg/storage/replica_command.go#L630)):*
> ```Go
> 	retryOpts := base.DefaultRetryOptions()
> 	retryOpts.MaxRetries = 5
> retryLoop:
> ```

Feel free to dismiss this suggestion, but I always try to avoid labels because they're hard to follow. Here it seems that you could just drop the `continue`below and add a check `if lastErr != nil { lastErr = nil; continue }` before the `return nil`. You don't need to propagate `lastErr` outside of the retry loop because at that point you know that all replicas caught the same error, so you can just return one you manufacture right at the end.
___
*[pkg/storage/replica_command.go, line 631 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPeTDbb9CxfSIMlcF9s:-LPeTDbb9CxfSIMlcF9t:b-z4lxyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7578c501755025d8fe4dff109e452871ca3e9432/pkg/storage/replica_command.go#L631)):*
> ```Go
> 	retryOpts.MaxRetries = 5
> retryLoop:
> 	for retrier := retry.StartWithCtx(ctx, retryOpts); retrier.Next(); {
> ```

Can we fail-fast if we're not the leader? Seems silly to spend a number of retries here and block merges that could actually proceed.
___
*[pkg/storage/replica_command.go, line 640 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPeTbX0EdMVhqIYKjhy:-LPeTbX0EdMVhqIYKjhz:b-nfjhtj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7578c501755025d8fe4dff109e452871ca3e9432/pkg/storage/replica_command.go#L640)):*
> ```Go
> 			// that the replicas are initialized, and the merge will (hopefully) be
> 			// retried on the leader.
> 			if raftStatus.Progress[uint64(r.ReplicaID)].Match == 0 {
> ```

`raftStatus` could be `nil`. Probably not in reality since you've already somehow made sure that it's initialized before getting here, but still.

**Review by bdarnell — APPROVED** [review]
:lgtm: with @tschottdorf 's comments

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-:-LPgCsBm90O3jeaXSX32:b-dslhib)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by benesch — COMMENTED** [review]
Tobi and I chatted and decided that it would be vastly simpler to just introduce a WaitForReplicaInit RPC into the stores server. Neither of us were fully confident that we wouldn't get stuck in a leader-but-not-leaseholder loop with the previous approach, and the code to transfer Raft leadership to match the leaseholdership is, uhm, hairy. (For example: do you want transferRaftLeader or transferRaftLeadership?)

PTAL!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-:-LPgPwuvEdqq9yEyinoK:b-6r3wr7)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/replica_command.go, line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPeTjB0-EH3WhpkwgVF:-LPgQDU48bqa01PAb-uB:b-cxchxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7578c501755025d8fe4dff109e452871ca3e9432/pkg/storage/replica_command.go#L630)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Feel free to dismiss this suggestion, but I always try to avoid labels because they're hard to follow. Here it seems that you could just drop the `continue`below and add a check `if lastErr != nil { lastErr = nil; continue }` before the `return nil`. You don't need to propagate `lastErr` outside of the retry loop because at that point you know that all replicas caught the same error, so you can just return one you manufacture right at the end.
</blockquote></details>

No longer relevant.
___
*[pkg/storage/replica_command.go, line 631 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPeTDbb9CxfSIMlcF9s:-LPgQAPvC6-NGHD-HWYs:b-cxchxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7578c501755025d8fe4dff109e452871ca3e9432/pkg/storage/replica_command.go#L631)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Can we fail-fast if we're not the leader? Seems silly to spend a number of retries here and block merges that could actually proceed.
</blockquote></details>

No longer relevant.
___
*[pkg/storage/replica_command.go, line 640 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPeTbX0EdMVhqIYKjhy:-LPgQETFC9En_8_UNogE:b-cxchxo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7578c501755025d8fe4dff109e452871ca3e9432/pkg/storage/replica_command.go#L640)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

`raftStatus` could be `nil`. Probably not in reality since you've already somehow made sure that it's initialized before getting here, but still.
</blockquote></details>

No longer relevant.

**Review by benesch — COMMENTED** [review]
I am curious to know if this has implications for backporting. I don't love the idea of introducing a new RPC in a patch release.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-:-LPgQj-Q9xmWM6ZV3xIo:b68x9ff)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Review by bdarnell — APPROVED** [review]
Adding this RPC in a patch release is fine since it'll only be used if the cluster setting is enabled, and that setting isn't changing in a patch release. Even if the user upgrades directly from 2.1.0 to 2.2.0 (which will flip the default), the loop that calls this RPC should just bail out and abort the merge if it gets an unknown RPC error.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-:-LPgaIKACLT-Y9iqIdgg:bd9re3t)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/storage/stores_server.go, line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPga7QX7aYSqD8zcM1D:-LPga7QX7aYSqD8zcM1E:b3vxpt7) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdf6bf5724ed41385f462edb9fb760a5f7f4baa9/pkg/storage/stores_server.go#L119)):*
> ```Go
> 	resp := &WaitForReplicaInitResponse{}
> 	err := is.execStoreCommand(req.StoreRequestHeader, func(s *Store) error {
> 		retryOpts := retry.Options{InitialBackoff: 10 * time.Millisecond}
> ```

There should be an upper bound on this retry loop. Better to return an error and abort the merge than to hang indefinitely.

**Comment by benesch** [comment]
> There should be an upper bound on this retry loop. Better to return an error and abort the merge than to hang indefinitely.

The upper bound is enforced by context timeouts. (This is how WaitForApplication works too.) The client sets the timeout at 5s currently. Is that sufficient or do you want me to add a server-side upper bound too?

**Review by tbg — APPROVED** [review]
:lgtm:

> Reviewed 5 of 5 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-:-LPjnGU0-6OywKByJumK:bnfp4nl)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/storage/replica_command.go, line 633 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPjmm5b4ToZK7f1n0m0:-LPjmm5b4ToZK7f1n0m1:b-k12oag) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdf6bf5724ed41385f462edb9fb760a5f7f4baa9/pkg/storage/replica_command.go#L633)):*
> ```Go
> 	g := ctxgroup.WithContext(ctx)
> 	for _, repl := range desc.Replicas {
> 		repl := repl
> ```

Things would get really awkward (and incorrect) without this line. I hate that this is necessary but that's the Go way. Please add a comment.
___
*[pkg/storage/stores_server.go, line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPga7QX7aYSqD8zcM1D:-LPjn5oZEm5_J9TXaE_k:b-ez564k) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdf6bf5724ed41385f462edb9fb760a5f7f4baa9/pkg/storage/stores_server.go#L119)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

There should be an upper bound on this retry loop. Better to return an error and abort the merge than to hang indefinitely. 
</blockquote></details>

Having the caller use context cancellation works for me, but add a comment here.

**Comment by bdarnell** [comment]
> The upper bound is enforced by context timeouts. (This is how WaitForApplication works too.) The client sets the timeout at 5s currently. Is that sufficient or do you want me to add a server-side upper bound too?

That's fine for now. I generally prefer limits set on both sides, but it doesn't need to change until we get around to doing a general review of all of our retry loops.

**Review by benesch — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31842)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

___
*[pkg/storage/replica_command.go, line 633 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPjmm5b4ToZK7f1n0m0:-LPmD4C35A-L0MLM0iMI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdf6bf5724ed41385f462edb9fb760a5f7f4baa9/pkg/storage/replica_command.go#L633)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Things would get really awkward (and incorrect) without this line. I hate that this is necessary but that's the Go way. Please add a comment.
</blockquote></details>

Done.
___
*[pkg/storage/stores_server.go, line 119 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-LPga7QX7aYSqD8zcM1D:-LPmD5498LiewO4XJ3XG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cdf6bf5724ed41385f462edb9fb760a5f7f4baa9/pkg/storage/stores_server.go#L119)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Having the caller use context cancellation works for me, but add a comment here.
</blockquote></details>

Done.

**Review by tbg — APPROVED** [review]
Wanna fix your imports and get this baby in?

> Reviewed 3 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31842#-:-LPqsK_NEOqbt0c4pNtR:b-x4gszr)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale)

**Comment by benesch** [comment]
bors r=tschottdorf,bdarnell

**Comment by benesch** [comment]
Whew, got it in for this night's test run. 🤞
